Adding callback to thingShadow.register

Torsph opened the pull request:
- Update the thing register function with a completion callback
- Add new test for the different states (ignoreDelta and persistentSubscribe)
- Updated README.md to reflect the new changes

Fixes issue #80 if I understood the aws and mqtt implementation correct. The new code waits for the mqtt lib to invoke the callback on _ignoreDeltas_ and _persistentSubscribe_ topics. When the mqtt lib has invoked both callbacks, will the register function invoke the register callback.

Maybe need for some more test cases, but please review and provide with feedback :)

rongsaws commented on thing/index.js:
```diff
@@ -477,15 +477,19 @@ function ThingShadowsClient(deviceOptions, thingShadowOptions) {
-            pending: false
+            pending: false,
+            pendingDeltaSubscribe: true,
+            pendingPersistentSubscribe: true
```
I think it's better not to treat delta ad other shadow topic subscriptions separately. One flag should be enough to tell if the shadow is ready to take actions. 

In the same vein, callback should also be invoked only once to be deterministic and avoid user confusion.

Torsph replied:
Agree. I was thinking about that. I will make the needed changes :)

rongsaws commented on thing/index.js:
```diff
@@ -515,7 +526,20 @@ function ThingShadowsClient(deviceOptions, thingShadowOptions) {
          if (thingShadows[thingName].persistentSubscribe === true) {
             this._handleSubscriptions(thingName, ['update', 'get', 'delete'], ['accepted', 'rejected'],
-               'subscribe');
+               'subscribe', function () {
```
subscribe callback should check granted subscriptions, see the underneath API here - https://github.com/mqttjs/MQTT.js#subscribe

If not all the topic subscriptions are granted, we could return an error in the register callback, and potentially fail following shadow operations.

Torsph replied:
OK. We should also clean up the subscriptions and remove the shadow. Good to add more error handling.

rongsaws commented on thing/index.js:
```diff
@@ -477,15 +477,19 @@ function ThingShadowsClient(deviceOptions, thingShadowOptions) {
             qos: 0,
-            pending: false
+            pending: false,
+            pendingDeltaSubscribe: true,
```
If this flag is pending, we should probably fail any shadow requests. We might be able to reuse the pending flag for the same purpose?

Torsph replied:
I guess we could. Will make a suggestion tomorrow or Monday.

Torsph commented on test/mock/mockMQTTClient.js:
```diff
@@ -18,6 +18,7 @@ function mockMQTTClient( wrapper, options ) {
         this.subscribeQosValues.length = 0;
         this.publishQosValues = new Array;
         this.publishQosValues.length = 0;
+        this.triggerError = null;
```
Maybe someone has a tip on how to force a error?

Torsph commented on test/thing-unit-tests.js:
```diff
@@ -69,6 +69,80 @@ describe( "thing shadow class unit tests", function() {
+
+        // Need local stub to trigger error, so just copy-paste the code
+        var localMockMqtt;
+        var fakeConnect = function (options) {
```
Needed the possibility to force an error to get the test working. So reused the code from the beforeEach.

rongsaws replied:
This approach works, another similar option is to replace triggerError with a function, and by default it returns empty result, but you can stub it using Sinon so it returns subscription grant result to be passed to callback. You could stub it in your test case so you don't need to have the code repetition here.

Torsph commented:
Will clean up the branch if it's accepted. How do you like it? rebase into one commit or something else?

rongsaws commented on thing/index.js:
```diff
@@ -526,6 +572,8 @@ function ThingShadowsClient(deviceOptions, thingShadowOptions) {
 
    this.unregister = function(thingName) {
       if (thingShadows.hasOwnProperty(thingName)) {
+         var topicsObject = [];
```
Use topicObjects to be consistent with your prior definition; and perhaps topicSpecs or toipcDefinitions is more readable?

Torsph replied:
Agree

rongsaws commented on thing/index.js:
```diff
@@ -514,9 +549,20 @@ function ThingShadowsClient(deviceOptions, thingShadowOptions) {
+         }
+
+         if (topicsObject.length > 0) {
+            this._handleSubscriptions(thingName, topicsObject, 'subscribe', callback);
```
From encapsulation point of view, pending should probably better be manipulated here, also you probably don't want to return granted etc in this register api, as whoever calls register() most likely won't care about what topics are accepted or not, what they do care is whether or not register() is successful. 

I would recommend to do something like this here

```
this._handleSubscriptions(thingName, topicsObject, 'subscribe', function(granted) {
        var res = "Not all subscriptions were granted";
        if (all topics are granted) {
               res = null;
               this.pending = false;  
        }
        if (!undefined(callback)) {
             callback(res);
        }
});
```


Also note _handleSubscriptions() does subscribe and unsubscribe, but for unsubscribe, there's no granted argument to the callback function.

Torsph replied:
Are you think about moving the logic for checking if all topics are granted into the callback? How are we going to handle all the other places that uses the _handleSubscriptions to sub?

rongsaws replied:
Checking granted code can be done in _handleSubscriptions() so users of it do need to check the ugly code 128 and so on. What I think it should pass to the callback is probably something like this:

callback(res, failedTopics) 

For successful subscriptions, both arguments would be null, for failed one, failed topics can be returned. For unsubscribe requests, failedTopics will always be empty.

rongsaws commented:
Hey, thanks again for working on this. 

Your second commit looks good overall. If you don't mind, I could pull it in to our internal pipeline and have a reviewed by 2nd pair of eyes after making some minor adjustments? We could then have it released after that.

Torsph commented:
Do you want me to fix the small issues before you pull it into the internal pipe?

rongsaws commented:
Certainly if you are up for it:)

Torsph commented:
Sure! Happy to help :)

Torsph commented on test/thing-unit-tests.js:
```diff
@@ -69,6 +69,70 @@ describe( "thing shadow class unit tests", function() {
+
+      it("should trigger error when a subscription fails", function () {
+
+        var stubTriggerError = sinon.stub(mockMQTTClient, 'triggerError', function(){return true;});
```
Haven't used sinon that much before, so don't know if I did it the correct way :)

Torsph commented:
- Changed the named to topicSpecs
- Moved the pending assignment
- The callback returns the error message and the topics that failed.
- Updated the test with sinon stub. (Haven't used sinon that much so it may be wrong, but it works :))

Hope that will do the work. Please just do all the changes needed in your internal review :)

Torsph commented:
Hi!

Do you have any update this PR? :)

jimisataws commented:
Hi Torsph,

Thanks for your contributions. These changes are included in our next release cycle and currently going through the internal pipeline.

Thanks
Jimi

liuszeng commented:
Addressed in release [v1.0.13](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v1.0.13).
